### PR TITLE
Class methods

### PR DESCRIPTION
**Pull request by a96lex**
### **What this PR does**

The most important thing is that almost all lower class methods (those of Script, Speech, Voice...) have been turned into classmethods in higher order classes. 

**Why is this helpful to our customers**

Basically, all class methods can be called directly on a class, instead of an instance. Using instances is also allowed, so this two snippets of code are now correct:

```python
import aflr
aflr.api_key = "your_key"
print(aflr.Voice().list())
```

```python
import aflr
aflr.api_key = "your_key"
print(aflr.Voice.list())
```

**Why is this helpful for us (sdk developers)**
With the previous architecture, a lot of code was repeated. Classes that shared similar methods (such as list, download, retrieve...) needed their own method for them.

Now we have added those methods in higher order classes that are then inherited by the end classes, those that users see and use. 

What this means effectively is that we can add a method to multiple classes by coding it only once and importing it to all the childs that need it.

All of the new higher order classes inherit APIRequest.

**Additional stuff**
Api key checker method in APIResource class:
 - Instead of checking the user's apiKey on the inizialization of the classes, we are doing it before building a header. The method is memoized as it will run on every api call.
 - Now we check for TypeErrors before ValueError. There was no way for a TypeError to appear before the ValueError even if the apiKey was not a string.
 - A small detail has been fixed: API key has 32 characters, not 30

co-authored by: @zeritte

### CODE REVIEW

File: aflr/api_request.py [hunk 1]
```diff
@@ -4,52 +4,72 @@
...
-        if self.api_key == None or self.api_key == "your-key":
-            self.api_key = os.environ.get("aflr_key", None)
-        if self.api_key == None or len(self.api_key) < 30:
+    @lru_cache(maxsize=10, typed=False)
+    def _api_key_checker(api_key=None):
+        if api_key == None or api_key == "your-key":
+            api_key = os.environ.get("aflr_key", None)
+        if not isinstance(api_key, str):
+            raise TypeError("api_key must be of type string.")
+        if api_key == None or len(api_key) < 32:
```
Comment (zeritte) [inline aflr/api_request.py:17]: no need to check if its None because it must be a string to reach here

File: aflr/api_request.py [hunk 2]
```diff
@@ -4,52 +4,72 @@
...
+    @classmethod
+    def _get_request(cls, url=None, path_param=None, request_params=None):
+        url = url or f"{aflr.api_base}{cls.resource_path}"
+        if hasattr(cls, "file_url"):
+            url = f"{aflr.api_base}{cls.file_url}"
+
+        if hasattr(cls, "bg_url"):
+            url = f"{aflr.api_base}{cls.bg_url}"
+
+        headers = cls._build_header()  # DRY. To be changed.
```
Comment (zeritte) [inline aflr/api_request.py:45]: I think you can remove the comment

File: aflr/helper_classes.py [hunk 3]
```diff
@@ -0,0 +1,70 @@
+from .api_request import APIRequest
+
+
+class ListableResource(APIRequest):
+    def __init__(self):
```
Comment (zeritte) [inline aflr/helper_classes.py:5]: is this init actually doing something?

File: aflr/helper_classes.py [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+    @classmethod
+    def list(cls, **args):
+        if args:
+            return cls._get_request(request_params=args)
+        else:
+            return cls._get_request()
+
+
+class CreatableResource(APIRequest):
+    def __init__(self):
```
Comment (zeritte) [inline aflr/helper_classes.py:17]: is this init actually doing something?

File: aflr/helper_classes.py [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+        super().__init__()
+
+    @classmethod
+    def create(cls, **params):
+        return cls._post_request(json=params)
+
+
+class RetrievableResource(APIRequest):
+    def __init__(self):
```
Comment (zeritte) [inline aflr/helper_classes.py:26]: is this init actually doing something?

File: aflr/helper_classes.py [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+            params.update({"section": section})
+
+        if hasattr(cls, "file_url"):
+            return cls._get_request(path_param=cls.file_url, request_params=params)
+        else:
+            return cls._get_request(request_params=params)
+
+
+class DownloadableResource(APIRequest):
+    def __init__(self):
```
Comment (zeritte) [inline aflr/helper_classes.py:44]: is this init actually doing something?

File: aflr/helper_classes.py [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+    @classmethod
+    def download(cls, scriptId, section=None, parameters=None, destination="."):
+        parameters = parameters or {}
+
+        try:
+            audio_files = cls.retrieve(scriptId, section, parameters)
+            audio_files.keys()
+        except Exception:
+            raise TypeError(
+                "Error retriving the audiofiles. The response object is not a dict"
```
Comment (zeritte) [inline aflr/helper_classes.py:56]: Error retrieving the audio files. Make sure you can retrieve them with the same parameters and try again.

is a better message I think

File: aflr/api_resources/voice.py [hunk 8]
```diff
@@ -1,18 +1,11 @@
-import aflr
-from aflr.api_request import APIRequest
+from aflr.helper_classes import ListableResource
 
 
-class Voice(APIRequest):
+class Voice(ListableResource):
     OBJECT_NAME = "voice"
+    resource_path = "/voice"
+    list_parameters_path = "voice/parameter"
```
Comment (zeritte) [inline aflr/api_resources/voice.py:7]: hmm why dont this have a backslash at the beginning

File: aflr/api_request.py [hunk 9]
```diff
@@ -4,52 +4,72 @@
...
         if request_params:
             r = requests.get(url=url, headers=headers, params=request_params)
         elif path_param:
-            r = requests.get(url=f"{url}/{path_param}", headers=headers)
+            r = requests.get(
+                url=f"{aflr.api_base}/{path_param}",
```
Comment (zeritte) [inline aflr/api_request.py:50]: shouldnt the url be url? you are doing some conditional checks above and deciding on the url

File: aflr/api_request.py [hunk 10]
```diff
@@ -4,52 +4,72 @@
...
-        if self.api_key == None or len(self.api_key) < 30:
+    @lru_cache(maxsize=10, typed=False)
+    def _api_key_checker(api_key=None):
+        if api_key == None or api_key == "your-key":
+            api_key = os.environ.get("aflr_key", None)
+        if not isinstance(api_key, str):
+            raise TypeError("api_key must be of type string.")
+        if api_key == None or len(api_key) < 32:
             raise ValueError(
                 "Please specify a valid api_key or create one here:\nhttps://console.api.audio"
```
Comment (zeritte) [inline aflr/api_request.py:19]: can we have a space after "\n" in case new line doesnt work

File: aflr/api_request.py [hunk 11]
```diff
@@ -4,52 +4,72 @@
...
 
 
 class APIRequest:
-    def __init__(self, api_key=None, api_base=None, **params):
-        self.api_key = api_key if api_key else aflr.api_key
-        self.api_base = api_base if api_base else aflr.api_base
-        if self.api_key == None or self.api_key == "your-key":
-            self.api_key = os.environ.get("aflr_key", None)
-        if self.api_key == None or len(self.api_key) < 30:
+    @lru_cache(maxsize=10, typed=False)
```
Comment (zeritte) [inline aflr/api_request.py:11]: why 10? why not a bigger number

File: aflr/api_request.py [hunk 12]
```diff
@@ -4,52 +4,72 @@
...
 
-    def _get_request(self, url, path_param=None, request_params=None):
-        headers = self._build_header()  # DRY. To be changed.
+    @classmethod
+    def _get_request(cls, url=None, path_param=None, request_params=None):
+        url = url or f"{aflr.api_base}{cls.resource_path}"
+        if hasattr(cls, "file_url"):
+            url = f"{aflr.api_base}{cls.file_url}"
+
+        if hasattr(cls, "bg_url"):
```
Comment (zeritte) [inline aflr/api_request.py:42]: what will happen if a class have both file and bg urls and I wanna get the file url?

**Review by zeritte — CHANGES_REQUESTED** [review]

**Comment by zeritte** [comment]
I think its a great work in general @a96lex. just needs a bit of more work & time. added some comments for you

File: aflr/api_request.py (hunk 1 above)
Comment (a96lex) [reply aflr/api_request.py:17]: removed

File: aflr/api_request.py (hunk 2 above)
Comment (a96lex) [reply aflr/api_request.py]: removed

File: aflr/helper_classes.py (hunk 4 above)
Comment (a96lex) [reply aflr/helper_classes.py]: removed all init methods for helper classes

File: aflr/helper_classes.py (hunk 7 above)
Comment (a96lex) [reply aflr/helper_classes.py:44]: changed

File: aflr/api_resources/voice.py (hunk 8 above)
Comment (a96lex) [reply aflr/api_resources/voice.py]: it has now

File: aflr/helper_classes.py [hunk 13]
```diff
@@ -0,0 +1,70 @@
...
+            )
+
+        if "url" in audio_files.keys():
+            local_filename = cls._download_request(
+                url=audio_files.get("url"), destination=destination
+            )
+            return local_filename
+        local_filenames = []
```
[the remark this answers is not in the document]
Comment (a96lex) [reply aflr/helper_classes.py:52]: added

File: aflr/api_request.py (hunk 9 above)
Comment (a96lex) [reply aflr/api_request.py]: yes it should. Fixed

File: aflr/api_request.py (hunk 10 above)
Comment (a96lex) [reply aflr/api_request.py:19]: sure

File: aflr/api_request.py (hunk 11 above)
Comment (a96lex) [reply aflr/api_request.py:11]: api key should not change often (or change at all normally). I think 10 is enough for cache

File: aflr/api_request.py (hunk 12 above)
Comment (a96lex) [reply aflr/api_request.py:40]: removed this logic, no more checkoing for file_url. Will need to rethink Sound class probably

File: aflr/api_request.py (hunk 11 above)
Comment (a96lex) [reply aflr/api_request.py]: removed memo

**Comment by springcoil** [comment]
Small nitpick - `Whi is this helpful for us (sdk developers)` should be `Why is this helpful for us (sdk developers)`

File: aflr/api_request.py [hunk 14]
```diff
@@ -7,49 +7,61 @@
...
-        if self.api_key == None or len(self.api_key) < 30:
+    def _api_key_checker(api_key=None):
+        if api_key == None or api_key == "your-key":
+            api_key = os.environ.get("aflr_key", None)
+        if not isinstance(api_key, str):
+            raise TypeError("api_key must be of type string.")
+        if len(api_key) < 32:
             raise ValueError(
-                "Please specify a valid api_key or create one here:\nhttps://console.api.audio"
+                "Please specify a valid api_key or create one here:\n https://console.api.audio"
```
Comment (springcoil) [inline aflr/api_request.py:17]: This is good. Nice that there's a space here

File: aflr/helper_classes.py [hunk 15]
```diff
@@ -0,0 +1,63 @@
+from .api_request import APIRequest
+
+
+class ListableResource(APIRequest):
```
Comment (springcoil) [inline aflr/helper_classes.py:4]: Is this spelt right? Is listable a word?

**Comment by springcoil** [comment]
@a96lex This is great work and quite a lot. I do feel though that one or two unit/integration tests would be useful here. To at least verify if everything is working.

File: aflr/helper_classes.py (hunk 15 above)
Comment (a96lex) [reply aflr/helper_classes.py:4]: it is a word, I checked.
Also, this nomenclature was inspired in stripe's one

**Comment by tonythree** [comment]
Hi Alex, great work overall. I really like the backwards compatibility and the abstract classes.
It would be cool to have the Readme file updated so people can save `()` in their calls.
I need to do a last test but I'd say we can merge this as soon as I test this.

File: aflr/api_request.py [hunk 16]
```diff
@@ -7,49 +7,61 @@
...
-        return {"x-api-key": self.api_key}
+        return {"x-api-key": aflr.api_key}
 
-    def _post_request(self, json, url=None):
-        url = self.url if not url else url
-        headers = self._build_header()
+    @classmethod
+    def _post_request(cls, json, url=None):
+        url = url or f"{aflr.api_base}{cls.resource_path}"
+        headers = cls._build_header()
```
Comment (zeritte) [inline aflr/api_request.py:29]: lets have new lines here as well like in the _get_request

File: aflr/helper_classes.py (hunk 15 above)
Comment (zeritte) [reply aflr/helper_classes.py:4]: it doesnt have to be a word tho, end users wont see it

File: aflr/api_request.py [hunk 17]
```diff
@@ -60,7 +72,9 @@ def _expanded_raise_for_status(self, res):
             res.raise_for_status()
         except HTTPError as e:
             if res.json():
-                raise HTTPError('{} \n Error Message from API: \n {}'.format(str(e), res.json()))
+                raise HTTPError(
+                    "{} \n Error Message from API: \n {}".format(str(e), res.json())
```
Comment (zeritte) [inline aflr/api_request.py:76]: can you change the order of these 2? because as I user I first check the latest part of the error

**Review by zeritte — APPROVED** [review]
looks great now! added a few comments. and why doesn't Readme include random call?

File: aflr/api_request.py (hunk 16 above)
Comment (a96lex) [reply aflr/api_request.py:34]: sure, added a convention :)

File: aflr/api_request.py (hunk 17 above)
Comment (a96lex) [reply aflr/api_request.py]: reordered

**Comment by a96lex** [comment]
> looks great now! added a few comments. and why doesn't Readme include random call?

@zeritte There is an open PR (#30) to modify this method. @UserName504 will add a usage example there

**Review by FeaR6077 — APPROVED** [review]
LGTM

**Review by tonythree — APPROVED** [review]
I reviewed everything and tested in local.
Works like a charm.

**Pull request merged**